modules/cms/classes/AutoDatasource.php

```diff
@@ -0,0 +1,510 @@
+<?php namespace Cms\Classes;
+
+use Cache;
+use Exception;
+use October\Rain\Halcyon\Model;
+use October\Rain\Halcyon\Processors\Processor;
+use October\Rain\Halcyon\Datasource\Datasource;
+use October\Rain\Halcyon\Exception\DeleteFileException;
+use October\Rain\Halcyon\Datasource\DatasourceInterface;
+
+/**
+ * Datasource that loads from other data sources automatically
+ *
+ * @package october\cms
+ * @author Luke Towers
+ */
+class AutoDatasource extends Datasource implements DatasourceInterface
+{
+    /**
+     * @var array The available datasource instances
+     */
+    protected $datasources = [];
+
+    /**
+     * @var array Local cache of paths available in the datasources
+     */
+    protected $pathCache = [];
+
+    /**
+     * @var boolean Flag on whether the cache should respect refresh requests
+     */
+    protected $allowCacheRefreshes = true;
+
+    /**
+     * @var string The key for the datasource to perform CRUD operations on
+     */
+    public $activeDatasourceKey = '';
+
+    /**
+     * Create a new datasource instance.
+     *
+     * @param array $datasources Array of datasources to utilize. Lower indexes = higher priority ['datasourceName' => $datasource]
+     * @return void
+     */
+    public function __construct(array $datasources)
+    {
+        $this->datasources = $datasources;
+
+        $this->activeDatasourceKey = array_keys($datasources)[0];
+
+        $this->populateCache();
+
+        $this->postProcessor = new Processor;
+    }
+
+    /**
+     * Populate the local cache of paths available in each datasource
+     *
+     * @param boolean $refresh Default false, set to true to force the cache to be rebuilt
+     * @return void
+     */
+    protected function populateCache($refresh = false)
+    {
+        $pathCache = [];
+        foreach ($this->datasources as $datasource) {
+            // Remove any existing cache data
+            if ($refresh && $this->allowCacheRefreshes) {
+                Cache::forget($datasource->getPathsCacheKey());
+            }
+
+            // Load the cache
+            $pathCache[] = Cache::rememberForever($datasource->getPathsCacheKey(), function () use ($datasource) {
+                return $datasource->getAvailablePaths();
+            });
+        }
+        $this->pathCache = $pathCache;
+    }
+
+    /**
+     * Check to see if the specified datasource has the provided Halcyon Model
+     *
+     * @param string $source The string key of the datasource to check
+     * @param Model $model The Halcyon Model to check for
+     * @return boolean
+     */
+    public function sourceHasModel(string $source, Model $model)
+    {
+        $result = false;
+
+        $sourcePaths = $this->getSourcePaths($source);
+
+        if (!empty($paths)) {
```

Comment:
I believe this is meant to be `$sourcePaths`.
```suggestion
        if (!empty($sourcePaths)) {
```